File: src/editor/musicCompatibility.ts
import type { ManagedAsset } from "./types";

export const LEGACY_OUTDOOR_MUSIC_BYTES = 60_224;
export const LEGACY_OUTDOOR_MUSIC_MD5 = "1A2E7CC637BCF082D21204E2DA1028B2";
export const LEGACY_OUTDOOR_MUSIC_SHA256 = "0ba9022f65d5cee0b57103c64264fc64f8fb0f84d63cf70445f2747ead9f2471";
export const OUTDOOR_MUSIC_REPLACEMENT_SHA256 = "f34da5b612972af0d6da8c000a1edc494a87b0c7aee627494a8d87f12132143f";
export const OUTDOOR_MUSIC_REPLACEMENT_BYTES = 62_054;
export const OUTDOOR_MUSIC_REPLACEMENT_PATH = "/bundled-libraries/providence/Outdoor%20Music.mod";

export function legacyOutdoorMusicSlot(name: string, bytes: number, sha256: string) {
  const match = /^Custom ([1-3]) Music$/i.exec(name.trim());
  if (!match || bytes !== LEGACY_OUTDOOR_MUSIC_BYTES || sha256.toLowerCase() !== LEGACY_OUTDOOR_MUSIC_SHA256) return null;
  return Number(match[1]);
}

export async function loadOutdoorMusicReplacement() {
  const response = await fetch(OUTDOOR_MUSIC_REPLACEMENT_PATH);
  if (!response.ok) throw new Error(`Bundled Outdoor Music replacement was not found (${response.status}).`);
  const bytes = new Uint8Array(await response.arrayBuffer());
  if (bytes.byteLength !== OUTDOOR_MUSIC_REPLACEMENT_BYTES) {
    throw new Error(`Bundled Outdoor Music replacement has ${bytes.byteLength} bytes; expected ${OUTDOOR_MUSIC_REPLACEMENT_BYTES}.`);
  }
  if (!await isOutdoorMusicReplacement(bytes)) {
    throw new Error("Bundled Outdoor Music replacement failed its immutable SHA-256 check.");
  }
  return bytes;
}

export async function isOutdoorMusicReplacement(bytes: Uint8Array) {
  if (bytes.byteLength !== OUTDOOR_MUSIC_REPLACEMENT_BYTES) return false;
  const buffer = new ArrayBuffer(bytes.byteLength);
  new Uint8Array(buffer).set(bytes);
  const digest = await crypto.subtle.digest("SHA-256", buffer);
  return Array.from(new Uint8Array(digest), (byte) => byte.toString(16).padStart(2, "0")).join("") === OUTDOOR_MUSIC_REPLACEMENT_SHA256;
}

export function legacyOutdoorMusicManagedAsset(slot: number, replacement: Uint8Array): ManagedAsset {
  const payload = `data:audio/x-mod;base64,${bytesToBase64(replacement)}`;
  return {
    id: `asset:legacy-outdoor-music:${slot}`,
    label: "Outdoor Music",
    kind: "music",
    resourceType: "MOD ",
    resourceId: slot,
    scenarioMusicSlot: slot,
    fileName: `Custom ${slot} Music`,
    originalPath: payload,
    previewPath: payload,
    resourcePath: payload,
    mimeType: "audio/x-mod",
    bytes: replacement.byteLength,
    sha256: OUTDOOR_MUSIC_REPLACEMENT_SHA256,
    width: null,
    height: null,
    durationMs: null,
    sampleRate: null,
    channels: null,
    exportState: "ready",
    libraryScope: "scenario",
    provenance: `legacy Outdoor Music compatibility alias (source ${LEGACY_OUTDOOR_MUSIC_BYTES} bytes, MD5 ${LEGACY_OUTDOOR_MUSIC_MD5})`,
    linkedEntity: `scenario-music:${slot}`,
    conversion: {
      target: "music",
      fitMode: null,
      scaleMode: null,
      matte: null,
      paletteMode: null,
      ditherMode: null,
      sourceWidth: null,
      sourceHeight: null,
      sourceDurationMs: null,
      sourceSampleRate: null,
      sourceChannels: null,
      finalWidth: null,
      finalHeight: null,
      warnings: ["The imported legacy MADG payload matched Realmz's known Outdoor Music fingerprint and was replaced with the bundled standard MOD compatibility version."]
    }
  };
}

function bytesToBase64(bytes: Uint8Array) {
  let binary = "";
  const chunkSize = 0x8000;
  for (let offset = 0; offset < bytes.length; offset += chunkSize) {
    binary += String.fromCharCode(...bytes.subarray(offset, offset + chunkSize));
  }
  return btoa(binary);
}

File: src-tauri/src/music_compatibility.rs
use sha2::{Digest, Sha256};

pub const LEGACY_OUTDOOR_MUSIC_BYTES: usize = 60_224;
pub const LEGACY_OUTDOOR_MUSIC_MD5: &str = "1A2E7CC637BCF082D21204E2DA1028B2";
pub const LEGACY_OUTDOOR_MUSIC_SHA256: &str =
    "0ba9022f65d5cee0b57103c64264fc64f8fb0f84d63cf70445f2747ead9f2471";
pub const OUTDOOR_MUSIC_REPLACEMENT_SHA256: &str =
    "f34da5b612972af0d6da8c000a1edc494a87b0c7aee627494a8d87f12132143f";
pub const OUTDOOR_MUSIC_REPLACEMENT: &[u8] =
    include_bytes!("../../public/bundled-libraries/providence/Outdoor Music.mod");

pub fn legacy_outdoor_music_slot(name: &str, bytes: &[u8]) -> Option<u8> {
    if !is_legacy_outdoor_music(bytes) {
        return None;
    }
    legacy_outdoor_music_slot_for_fingerprint(name, bytes.len(), LEGACY_OUTDOOR_MUSIC_SHA256)
}

pub fn legacy_outdoor_music_slot_for_fingerprint(
    name: &str,
    byte_length: usize,
    sha256: &str,
) -> Option<u8> {
    let slot = match name.trim().to_ascii_lowercase().as_str() {
        "custom 1 music" => 1,
        "custom 2 music" => 2,
        "custom 3 music" => 3,
        _ => return None,
    };
    (byte_length == LEGACY_OUTDOOR_MUSIC_BYTES
        && sha256.eq_ignore_ascii_case(LEGACY_OUTDOOR_MUSIC_SHA256))
    .then_some(slot)
}

pub fn is_legacy_outdoor_music(bytes: &[u8]) -> bool {
    bytes.len() == LEGACY_OUTDOOR_MUSIC_BYTES && sha256_hex(bytes) == LEGACY_OUTDOOR_MUSIC_SHA256
}

pub fn replacement_bytes() -> &'static [u8] {
    debug_assert_eq!(
        sha256_hex(OUTDOOR_MUSIC_REPLACEMENT),
        OUTDOOR_MUSIC_REPLACEMENT_SHA256
    );
    OUTDOOR_MUSIC_REPLACEMENT
}

pub fn is_outdoor_music_replacement(bytes: &[u8]) -> bool {
    bytes.len() == OUTDOOR_MUSIC_REPLACEMENT.len()
        && sha256_hex(bytes) == OUTDOOR_MUSIC_REPLACEMENT_SHA256
}

fn sha256_hex(bytes: &[u8]) -> String {
    format!("{:x}", Sha256::digest(bytes))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn bundled_replacement_is_a_different_standard_mod() {
        assert_eq!(OUTDOOR_MUSIC_REPLACEMENT.len(), 62_054);
        assert_eq!(&OUTDOOR_MUSIC_REPLACEMENT[1080..1084], b"M.K.");
        assert_eq!(
            sha256_hex(OUTDOOR_MUSIC_REPLACEMENT),
            OUTDOOR_MUSIC_REPLACEMENT_SHA256
        );
        assert!(!is_legacy_outdoor_music(OUTDOOR_MUSIC_REPLACEMENT));
    }

    #[test]
    fn legacy_alias_requires_the_exact_known_fingerprint_and_classic_slot_name() {
        assert_eq!(
            legacy_outdoor_music_slot_for_fingerprint(
                "Custom 2 Music",
                LEGACY_OUTDOOR_MUSIC_BYTES,
                LEGACY_OUTDOOR_MUSIC_SHA256,
            ),
            Some(2)
        );
        assert_eq!(
            legacy_outdoor_music_slot_for_fingerprint(
                "Custom 2 Music",
                LEGACY_OUTDOOR_MUSIC_BYTES - 1,
                LEGACY_OUTDOOR_MUSIC_SHA256,
            ),
            None
        );
        assert_eq!(
            legacy_outdoor_music_slot_for_fingerprint(
                "Outdoor Music",
                LEGACY_OUTDOOR_MUSIC_BYTES,
                LEGACY_OUTDOOR_MUSIC_SHA256,
            ),
            None
        );
    }
}
